File: oolab/src/main/java/agh/ics/oop/model/MapDirection.java
package agh.ics.oop.model;

public enum MapDirection {
    NORTH, NORTHEAST, EAST, SOUTHEAST, SOUTH, SOUTHWEST, WEST, NORTHWEST;

    private static final MapDirection[] directions = values();

    public MapDirection add(int i) {
        if (i < 0 || i > 7)
            throw new IllegalArgumentException("Turn should be an integer in [0,7].");
        return directions[(ordinal() + i) % directions.length];
    }

    public Vector2d toUnitVector() {
        return switch (this) {
            case NORTH -> new Vector2d(0, 1);
            case NORTHEAST -> new Vector2d(1, 1);
            case EAST -> new Vector2d(1, 0);
            case SOUTHEAST -> new Vector2d(1, -1);
            case SOUTH -> new Vector2d(0, -1);
            case SOUTHWEST -> new Vector2d(-1, -1);
            case WEST -> new Vector2d(-1, 0);
            case NORTHWEST -> new Vector2d(-1, 1);
        };
    }

    @Override
    public String toString() {
        return switch (this) {
            case NORTH -> "↑";
            case NORTHEAST -> "↗";
            case EAST -> "→";
            case SOUTHEAST -> "↘";
            case SOUTH -> "↓";
            case SOUTHWEST -> "↙";
            case WEST -> "←";
            case NORTHWEST -> "↖";
        };
    }


    public MapDirection reverse() {
        return switch (this) {
            case NORTHEAST -> SOUTHEAST;
            case SOUTHEAST -> NORTHEAST;
            case NORTHWEST -> SOUTHWEST;
            case SOUTHWEST -> NORTHWEST;
            case SOUTH->NORTH;
            case NORTH->SOUTH;
            case EAST -> EAST;
            case WEST -> WEST;
        };
    }

}

File: kotlinlab/src/main/kotlin/model/MapDirection.kt
package model

enum class MapDirection {
    NORTH, WEST, SOUTH, EAST;

    fun next(): MapDirection = entries[(ordinal - 1 + entries.size) % entries.size]

    fun previous(): MapDirection = entries[(ordinal + 1) % entries.size]

    fun toUnitVector(): Vector2d = when (this) {
        NORTH -> Vector2d(0, 1)
        WEST -> Vector2d(-1, 0)
        SOUTH -> Vector2d(0, -1)
        EAST -> Vector2d(1, 0)
    }

    override fun toString() = when (this) {
        NORTH -> "Północ"
        WEST -> "Zachód"
        SOUTH -> "Południe"
        EAST -> "Wschód"
    }
}
